Add glossary section to documentation (#1583)
FerretDB/FerretDB: website/docs/reference/glossary.md

---
sidebar_position: 2
---

# Glossary

## List of FerretDB terminologies

*This section contains a list of common terminologies related to FerretDB*.

### A

#### aggregation

> A group of operations used to reduce or merge large data sets.

Reviewer: In principle, data sets don't have to be large. It can be executed with small data sets too. The idea is that during aggregation, we can execute some special operations - like compute some statistics, or some math operators etc. Aggregation can be split into a few stages, e.g. first we filter the necessary data and then execute some statistics calculation for this data. But such verbal description is not for the glossary I guess :)

So, overall, I'd say the idea of aggregation is that we process documents and apply various operations to them.

See [list of supported aggregation operations and commands here](./supported_commands.md#aggregation-pipelines).

#### aggregation pipeline

A set of operators that lets you perform complex operations that aggregate and summarize values.
See [list of supported aggregation pipeline operators](./supported_commands.md#aggregation-pipeline-operators) here.

---

### B

#### Beacon

The telemetry service of FerretDB.
See [telemetry](../telemetry.md) for more details.

#### BSON

BSON is a serialized binary file format for storing JSON-like documents.

#### BSON types

The list of types that the BSON format supports.
BSON offers support for additional data types compared to JSON, such as `timestamp`, `date`, `ObjectId`, and `binary`.

---

### C

#### collection

A group of documents in a non-relational database.
It is comparable to a table in a relational database.

#### CRUD

The four basic operations of a database: Create, Read, Update, and Delete.
See [Basic FerretDB CRUD operations here](../basic_operations/index.md).

---

### D

#### database

An organized repository for collections containing its own sets of documents, and data.

> #### database command

Reviewer: Hmm, do we separate "database commands" and "collection commands"? Or do we just call everything a "database command"? Or just a "command"?

Author: I believe database commands cover for everything in this regard.

The set of commands in FerretDB.
For more information, see [supported commands](./supported_commands.md) for more details.

#### document

A record in a collection that comprises key-value pairs.
See [Documents](../understanding_ferretdb.md#documents) for more.

#### dot notation

Dot notation is used to reference or access the elements in an array or in an embedded document.
See [dot notation](../understanding_ferretdb.md#dot-notation) for more details.

---

### F

#### field

Similar to columns in a relational database.
They are represented as field name-value pairs and describe the kind of data in a document.

---

### I

#### index

A data structure used for identifying and querying records in a collection.
It helps to limit the number of documents to search through or inspect in a collection.

> Examples include unique `_id` index, user-defined index, hashed index, and partial index.

Reviewer: Hmm... Unique indexes are not necessary for `_id` only.

Author: Sorry, I want to understand: "Unique" should apply to all the indexes, right?

Reviewer: Yes, yes. I mean, in principle, any field can be indexed as unique. This is not what we offer at FerretDB at the moment (we don't offer it for `_id` either).

Reviewer: And the index for `_id` should be created automatically when collection is created, so this is the most straightforward unique index, yes ;)

Author: Great. Thanks. I'll tweak the definition a bit to fit that.

---

### J

#### JSON

An acronym for JavaScript Object Notation.
It is a structured data format with human-readable text to store data objects composed of attribute-value pairs.

#### JSONB

JSONB is a data type of PostgreSQL that stores JSON data as a decomposed binary format.

#### ObjectId

A defining 12-byte type that ensures singularity and uniques within a collection and are used to represent the default values for the `_id` fields.

> #### operator

Reviewer: How do we call things like `insert`, `update`, `find`, `count` etc? Methods? Operations? Should we add a line about those things to our glossary?

Author: Good question. They should be operations, and in some instances,  commands. I'm no longer sure if method is appropriate. Let me know what you think.

Reviewer: I'm not sure either :) I think, currently, quite often, we use "methods" or "commands" as synonyms, and I think this is one of the reasons why we want to have a glossary - to know how to call them correctly. I see that in the code we call things like "insert", "find" etc commands - https://github.com/FerretDB/FerretDB/blob/7c94741baebed1cb24040a9a712353a1f174a394/internal/handlers/common/msg_listcommands.go#L42. So, maybe "command" is a good word.

Author: Let's use "commands" for them going forward.

---
sidebar_position: 2
---

# Glossary

## List of FerretDB terminologies

*This section contains a list of common terminologies related to FerretDB*.

### A

#### aggregation

A group of operations used to reduce or summarize large data sets.
See [list of supported aggregation operations and commands here](./supported_commands.md#aggregation-pipelines).

#### aggregation pipeline

A set of operators that lets you perform complex operations that aggregate and summarize values.
See [list of supported aggregation pipeline operators](./supported_commands.md#aggregation-pipeline-operators) here.

---

### B

#### Beacon

> The telemetry service of FerretDB.

Reviewer 2: Should we link to the beacon.ferretdb.io or telemetry page?

Author: Probably the telemetry page for now. Once the beacon page is ready, we can update the link.

Reviewer 2: The telemetry page is ready :)

#### BSON

> Stands for “binary JSON”, which is a serialized binary file format for storing JSON-like documents.

Reviewer 3: I think we don't need to mention that it stands for `binary JSON` here. 
```suggestion
BSON is a serialized binary file format for storing JSON-like documents.
```

#### BSON types

The list of types that the BSON format supports.
BSON offers support for additional data types compared to JSON, such as `timestamp`, `date`, `ObjectId`, and `binary`.

---

### C

#### collection

A group of documents in a non-relational database.
It is comparable to a table in a relational database.

#### CRUD

The four basic operations of a database: Create, Read, Update, and Delete.
See [Basic FerretDB CRUD operations here](../basic_operations/index.md).

---

### D

> #### dance

Reviewer 2: That should not be user-visible anywhere

The integration testing tool of FerretDB, which is named after the Ferret war dance.
See [dance repository](https://github.com/FerretDB/dance) for more details.

#### database

An organized repository for collections containing its own sets of documents, and data.

#### database command

The set of commands in FerretDB.
For more information, see [supported commands](./supported_commands.md) for more details.

#### document

A record in a collection that comprises key-value pairs.
See [Documents](../understanding_ferretdb.md#documents) for more.

#### dot notation

Dot notation is used to reference or access the elements in an array or in an embedded document.
See [dot notation](../understanding_ferretdb.md#dot-notation) for more details.

---

### F

#### field

Similar to columns in a relational database.
They are represented as field-value pairs and describe the kind of data in a document.

---

### G

> #### github-actions

Reviewer 2: That should not be user-visible anywhere

A repository that holds the shared GitHub Actions across all FerretDB repositories.
See [github-actions repository](https://github.com/FerretDB/github-actions) for more details.

---

### I

#### index

> A data structure used for identifying and querying records in a database.

Reviewer 3: Hmm, I'm not sure what we are referring here. Is it index of the dot notation in the array?

---

### J

#### JSON

An acronym for JavaScript Object Notation.

> It is a structured data format with human-readable text  to store data objects composed of attribute-value pairs.

Reviewer 3: ```suggestion
It is a structured data format with human-readable text to store data objects composed of attribute-value pairs.
```

#### JSONB

> JSONB is a data type that stores JSON data as a decomposed binary format.

Reviewer 2: It should mention PostgreSQL

> JSONB is a data type that stores JSON data as a decomposed binary format.

Reviewer 3: ```suggestion
JSONB is a data type of PostgreSQL that stores JSON data as a decomposed binary format.
```

#### ObjectId

A defining 12-byte type that ensures singularity and uniques within a collection and are used to represent the default values for the `_id` fields.

> It contains a 4-byte timestamp value measured in seconds, uniquely random machine and process-generated ID, and an incremental counter.

Reviewer 3: That is how we generate ObjectId. However ObjectId if specified by a user can be anything except a few exception one of which is nil. Let's double check if anything else is exception.

Reviewer 2: Not even "we", but the driver they use. It might be better just to remove that sentence

Reviewer 4: > ObjectId if specified by a user can be anything
That's not quite true. ObjectId can be only 12-byte type described in this paragraph. `_id` field can have any type except regex and array.

Reviewer 2: Anything that is 12-bytes long :)

#### operator

A keyword that starts with a `$` character to query, update, or transform data.

---

### P

> #### PJSON

Reviewer 2: Probably, not user-visible

Author: Included this because we have an article on it here: https://www.ferretdb.io/pjson-how-to-store-bson-in-jsonb/. But let me know what you think.

FerretDB’s mapping system which converts BSON into JSONB.
It uses embedded data in BSON containing special keys that starts with `$` along with information about the data types and field order to serialize data into JSONB.
See [this article for more details on PJSON](https://www.ferretdb.io/pjson-how-to-store-bson-in-jsonb/).

#### primary key

An immutable identifier for a record.
The primary key of a documents is stored in the `_id` field, which typically contains the `ObjectId`.

#### PostgreSQL

An open source relational database.
FerretDB uses PostgreSQL as a database engine.

---

### T

#### Tigris

> Also known as Tigris Data.

Reviewer 2: Not really – Tigris is a product name, Tigris Data Inc. is company name
